app/src/main/java/nerd/tuxmobil/fahrplan/congress/details/SessionDetailsViewModel.kt

```diff
@@ -133,60 +69,140 @@ class SessionDetailsViewModel @JvmOverloads constructor(
+    val setAlarm = SingleLiveEvent<Unit>()
+    val navigateToRoom = SingleLiveEvent<Uri>()
+    val closeDetails = SingleLiveEvent<Unit>()
+
+    private fun Session.toSelectedSessionParameter(): SelectedSessionParameter {
+        val useDeviceTimeZone = repository.readUseDeviceTimeZoneEnabled()
+        val formattedZonedDateTime = formattingDelegate.getFormattedDateTimeShort(useDeviceTimeZone, dateUTC, timeZoneOffset)
+        val formattedAbstract = markdownConversion.markdownLinksToHtmlLinks(abstractt)
+        val formattedDescription = markdownConversion.markdownLinksToHtmlLinks(description)
+        // language=regex
+        val linksHtml = getLinks().replace("\\),".toRegex(), ")<br>")
+        val formattedLinks = markdownConversion.markdownLinksToHtmlLinks(linksHtml)
+        val sessionUrl = sessionUrlComposition.getSessionUrl(this)
+        val sessionLink = if (sessionUrl.isEmpty()) "" else "<a href=\"$sessionUrl\">$sessionUrl</a>"
+        val isFeedbackUrlEmpty = feedbackUrlComposer.getFeedbackUrl(this).isEmpty()
+        val isC3NavRoomNameEmpty = roomForC3NavConverter.convert(room).isEmpty()
+        return SelectedSessionParameter(
+            // Details content
+            hasDateUtc = dateUTC > 0,
+            formattedZonedDateTime = formattedZonedDateTime,
+            roomName = room.orEmpty(),
+            sessionId = sessionId,
+            title = title.orEmpty(),
+            subtitle = subtitle.orEmpty(),
+            speakerNames = formattedSpeakers,
+            abstract = abstractt.orEmpty(),
+            formattedAbstract = formattedAbstract,
+            description = description.orEmpty(),
+            formattedDescription = formattedDescription,
+            hasLinks = getLinks().isNotEmpty(),
+            formattedLinks = formattedLinks,
+            hasWikiLinks = getLinks().containsWikiLink(),
+            sessionLink = sessionLink,
+            // Options menu
+            isFlaggedAsFavorite = highlight,
+            hasAlarm = hasAlarm,
+            isFeedbackUrlEmpty = isFeedbackUrlEmpty,
+            isC3NavRoomNameEmpty = isC3NavRoomNameEmpty
+        )
+    }
+
+    fun openFeedback() {
+        viewModelScope.launch(executionContext.database) {
+            val session = repository.loadSelectedSession()
+            val uri = feedbackUrlComposer.getFeedbackUrl(session).toUri()
+            openFeedBack.postValue(uri)
         }
-        R.id.menu_item_share_session_json -> {
-            val formattedSessions = session.toJson()
-            viewActionHandler.shareAsJson(formattedSessions)
-            true
+    }
+
+    fun share() {
+        viewModelScope.launch(executionContext.database) {
+            val session = repository.loadSelectedSession()
+            val timeZoneId = repository.readMeta().timeZoneId
+            simpleSessionFormat.format(session, timeZoneId).let { formattedSession ->
+                shareSimple.postValue(formattedSession)
+            }
         }
```

Comment:
I am wondering if a custom `launchWithDataBaseAndSession` could make this and the following methods less repetitive and the code a bit lighter, since they follow all the same pattern (launch -> loadSession -> do smt).

Reply:
:heavy_check_mark: Extracted in d6583a5dd41c1ab01cb82979b18e28a5d1e52546.